### ISSUE TITLE
`IndexError` in `shapely.geometry.shape` when parsing a MultiPolygon with some empty members

### ISSUE THREAD

**Issue by TomAugspurger** [shapely/shapely#1407]
## Expected behavior and actual behavior.

Create a `MultiPolygon`.

## Steps to reproduce the problem.

I think this is valid GeoJSON, but shapely errors when trying to convert it.

```python
import shapely.geometry
data = {
    "type": "MultiPolygon",
    "coordinates": [
        [],
        [[[0, 0],
          [1, 0],
          [1, 1],
          [0, 1],
          [0, 0]]]
    ]
}

shapely.geometry.shape(data)
```

That raises with

```pytb
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Input In [3], in <cell line: 13>()
      1 data = {
      2     "type": "MultiPolygon",
      3     "coordinates": [
   (...)
     10     ]
     11 }
---> 13 shapely.geometry.shape(data)

File ~/src/stac-utils/stac-geoparquet/.venv/lib/python3.8/site-packages/shapely/geometry/geo.py:118, in shape(context)
    116     return MultiLineString(ob["coordinates"])
    117 elif geom_type == "multipolygon":
--> 118     return MultiPolygon(ob["coordinates"], context_type='geojson')
    119 elif geom_type == "geometrycollection":
    120     geoms = [shape(g) for g in ob.get("geometries", [])]

File ~/src/stac-utils/stac-geoparquet/.venv/lib/python3.8/site-packages/shapely/geometry/multipolygon.py:62, in MultiPolygon.__init__(self, polygons, context_type)
     60     geom, n = geos_multipolygon_from_polygons(polygons)
     61 elif context_type == 'geojson':
---> 62     geom, n = geos_multipolygon_from_py(polygons)
     63 self._set_geom(geom)
     64 self._ndim = n

File ~/src/stac-utils/stac-geoparquet/.venv/lib/python3.8/site-packages/shapely/geometry/multipolygon.py:146, in geos_multipolygon_from_py(ob)
    143 L = len(ob)
    144 assert L >= 1
--> 146 N = len(ob[0][0][0])
    147 assert N == 2 or N == 3
    149 subs = (c_void_p * L)()

IndexError: list index out of range
```

## Operating system

Linux

## Shapely version and provenance

1.8.2 from PyPI

**Comment by Georgea75** [comment]
Hi,

I am interested in helping out with this and I am trying to understand what the expected behavior is around handling empty components in multi-geometries.  

I ran a couple of tests with GeoJSON inputs containing one empty geometry and one valid one.

MultiPoint raises a value error
```
In [1]: import shapely.geometry

In [2]: point = {
   ...:   "type": "MultiPoint",
   ...:   "coordinates": [
   ...:     [],
   ...:     [120.5, -35.2]
   ...:   ]
   ...: }

In [3]: shapely.geometry.shape(point)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[3], line 1
----> 1 shapely.geometry.shape(point)

File ~/shapely_test/venv/lib/python3.12/site-packages/shapely/geometry/geo.py:109, in shape(context)
    107     return Polygon(ob["coordinates"][0], ob["coordinates"][1:])
    108 elif geom_type == "multipoint":
--> 109     return MultiPoint(ob["coordinates"])
    110 elif geom_type == "multilinestring":
    111     return MultiLineString(ob["coordinates"])

File ~/shapely_test/venv/lib/python3.12/site-packages/shapely/geometry/multipoint.py:64, in MultiPoint.__new__(self, points)
     62 subs = []
     63 for item in points:
---> 64     p = point.Point(item)
     65     if p.is_empty:
     66         raise EmptyPartError("Can't create MultiPoint with empty component")

File ~/shapely_test/venv/lib/python3.12/site-packages/shapely/geometry/point.py:81, in Point.__new__(self, *args)
     79 if not np.issubdtype(coords.dtype, np.number):
     80     coords = [float(c) for c in coords]
---> 81 geom = shapely.points(coords)
     82 if not isinstance(geom, Point):
     83     raise ValueError("Invalid values passed to Point constructor")

File ~/shapely_test/venv/lib/python3.12/site-packages/shapely/decorators.py:173, in deprecate_positional.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
    171 @wraps(func)
    172 def wrapper(*args, **kwargs):
--> 173     result = func(*args, **kwargs)
    175     n = len(args)
    176     if n > warn_from:

File ~/shapely_test/venv/lib/python3.12/site-packages/shapely/decorators.py:88, in multithreading_enabled.<locals>.wrapped(*args, **kwargs)
     86     for arr in array_args:
     87         arr.flags.writeable = False
---> 88     return func(*args, **kwargs)
     89 finally:
     90     for arr, old_flag in zip(array_args, old_flags):

File ~/shapely_test/venv/lib/python3.12/site-packages/shapely/creation.py:124, in points(coords, y, z, indices, handle_nan, out, **kwargs)
    122     handle_nan = HandleNaN.get_value(handle_nan)
    123 if indices is None:
--> 124     return lib.points(coords, np.intc(handle_nan), out=out, **kwargs)
    125 else:
    126     return simple_geometries_1d(
    127         coords, indices, GeometryType.POINT, handle_nan=handle_nan, out=out
    128     )

ValueError: The ordinate (last) dimension should be 2 or 3, got 0
```

Where MultiLineString raises a EmptyPartError

```
In [1]: import shapely.geometry

In [2]: line_string = {
   ...:   "type": "MultiLineString",
   ...:   "coordinates": [
   ...:     [],
   ...:     [
   ...:       [10.0, 10.0],
   ...:       [20.0, 20.0]
   ...:     ]
   ...:   ]
   ...: }

In [3]: shapely.geometry.shape(line_string)
---------------------------------------------------------------------------
EmptyPartError                            Traceback (most recent call last)
Cell In[3], line 1
----> 1 shapely.geometry.shape(line_string)

File ~/shapely_test/venv/lib/python3.12/site-packages/shapely/geometry/geo.py:111, in shape(context)
    109     return MultiPoint(ob["coordinates"])
    110 elif geom_type == "multilinestring":
--> 111     return MultiLineString(ob["coordinates"])
    112 elif geom_type == "multipolygon":
    113     return MultiPolygon([[c[0], c[1:]] for c in ob["coordinates"]])

File ~/shapely_test/venv/lib/python3.12/site-packages/shapely/geometry/multilinestring.py:52, in MultiLineString.__new__(self, lines)
     50     line = linestring.LineString(item)
     51     if line.is_empty:
---> 52         raise EmptyPartError(
     53             "Can't create MultiLineString with empty component"
     54         )
     55     subs.append(line)
     57 if len(lines) == 0:

EmptyPartError: Can't create MultiLineString with empty component
```
Should Shapely allow empty components in multi-geometries, or should it consistently raise `EmptyPartError` across all multi-geometry types?

**Comment by jorisvandenbossche** [comment]
@Georgea75 late response, but that is a good question..

And I don't think I have a definitive answer for it. 
As you show, at least in certain cases, we currently have the explicit check (and thus intentional behaviour, not an accidental error) for disallowing empty parts, like for `MultiLineString` (and `MultiPoint`). And so one argument is that we should at least make this consistent with the other geometry types' constructors. 

However, this behaviour comes from the class constructors itself. For example, that MultiLineString example constructing it from two linestrings, this also errors:

```python
>>> shapely.MultiLineString([shapely.LineString(), shapely.LineString([(0, 1), (2, 3)])])
...
EmptyPartError: Can't create MultiLineString with empty component
```

But at that level, this actually does work right now for MultiPolygon (it seems to skip the empty one, though):

```python
>>> shapely.MultiPolygon([shapely.Polygon(), shapely.Polygon([(0, 0), (1, 0), (0, 1)])])
<MULTIPOLYGON (((0 0, 1 0, 0 1, 0 0)))>
```

(and since it works, I think we should probably keep that working, and not start raising an EmptyPartError)

Also, on the other hand, the `shape()` constructor is meant for creating a shapely Geometry from GeoJSON-like data / `__geo_interface__`. And for example, our GeoJSON reader does allow such empty parts. Taking the example from the top post:

```python
>>> import json
>>> json.dumps(data)
'{"type": "MultiPolygon", "coordinates": [[], [[[0, 0], [1, 0], [1, 1], [0, 1], [0, 0]]]]}'
>>> shapely.from_geojson(json.dumps(data))
<MULTIPOLYGON (EMPTY, ((0 0, 1 0, 1 1, 0 1, 0 0)))>
```

And roundtripping this multi-part geometry with an empty part with `__geo_interface__` / `shapely.geometry.shape` actually does seem to work:

```python
>>> geom = shapely.from_geojson(json.dumps(data))
>>> geom.__geo_interface__
{'type': 'MultiPolygon',
 'coordinates': [((),),
  (((0.0, 0.0), (1.0, 0.0), (1.0, 1.0), (0.0, 1.0), (0.0, 0.0)),)]}
>>> shapely.geometry.shape(geom.__geo_interface__)
<MULTIPOLYGON (EMPTY, ((0 0, 1 0, 1 1, 0 1, 0 0)))>
```

It does work for MultiPolygon, since that constructor doesn't disallow empty parts (and can handle the empty nested `((), )` in contrast to the `[]` from the top post) 
But trying such a roundtrip for a MultiLineString then actually gives the EmptyPartError.

In summary, I think a general guideline to design those constructors could be that a `__geo_interface__` / `shape()` roundtrip should work.
